main/src/library/Reducible.flix

```diff
@@ -0,0 +1,183 @@
+///
+/// This type class is heavily inspired by the Cats implementation of Reducible:
+/// https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/Reducible.scala
+///
+pub lawless class Reducible[t: Type -> Type] {
+
+    ///
+    /// Left-associative reduction of a structure.
+    /// Applies `g` to the initial element of `t` and combines it
+    /// with the remainder of `t` using `f` going from left to right.
+    ///
+    pub def reduceLeftTo(f: (b, a) -> b & ef1, g: a -> b & ef2, t: t[a]): b & (ef1 and ef2)
+
+    ///
+    /// Right-associative reduction of a structure.
+    /// Applies `g` to the initial element of `t` and combines it
+    /// with the remainder of `t` using `f` going from right to left.
+    ///
+    pub def reduceRightTo(f: (a, b) -> b & ef1, g: a -> b & ef2, t: t[a]): b & (ef1 and ef2)
+
+    ///
+    /// Left-associative reduction on `t` using `f`.
+    ///
+    pub def reduceLeft(f: (a, a) -> a & ef, t: t[a]): a & ef =
+        Reducible.reduceLeftTo(f, identity, t)
+
+    ///
+    /// Right-associative reduction on `t` using `f`.
+    ///
+    pub def reduceRight(f: (a, a) -> a & ef, t: t[a]): a & ef =
+        Reducible.reduceRightTo(f, identity, t)
+
+    ///
+    /// Reduce `t` using the derived `SemiGroup` instance.
+    ///
+    pub def reduce(t: t[a]): a with SemiGroup[a] =
+        Reducible.reduceLeft((acc, a) -> SemiGroup.combine(acc, a), t)
+
+    ///
+    /// Applies `f` to each element of `t` and combines them using the derived `SemiGroup` instance.
+    ///
+    pub def reduceMap(f: a -> b & ef, t: t[a]): b & ef with SemiGroup[b] =
+        Reducible.reduceLeftTo((b, a) -> SemiGroup.combine(b, f(a)), f, t)
+
+    ///
+    /// Left-associative fold of a structure.
+    /// Applies `f` to a start value `s` and all elements in `t` going from left to right.
+    ///
+    pub def foldLeft(f: (b, a) -> b & ef, s: b, t: t[a]): b & ef =
+        Reducible.reduceLeftTo(f, a -> f(s, a), t)
+
+    ///
+    /// Right-associative fold of a structure.
+    /// Applies `f` to a start value `s` and all elements in `t` going from right to left.
+    ///
+    pub def foldRight(f: (a, b) -> b & ef, s: b, t: t[a]): b & ef =
+        Reducible.reduceRightTo(f, a -> f(a, s), t)
+
+    ///
+    /// Returns the first element of `t`.
+    ///
+    pub def head(t: t[a]): a =
+        Reducible.reduceLeft((acc, _) -> acc, t)
+
+    ///
+    /// Returns the last element of `t`.
+    ///
+    pub def last(t: t[a]): a =
+        Reducible.reduceRight((_, acc) -> acc, t) // Could use reduceLeft w. lambda (_, a) -> a
+
+    ///
+    /// Returns the number of elements in `t` that satisfy the predicate `f`.
+    ///
+    pub def count(f: a -> Bool, t: t[a]): Int32 =
+        Reducible.reduceMap(a -> if (f(a)) 1 else 0, t)
+
+    ///
+    /// Returns the number of elements in `t`.
+    ///
+    pub def length(t: t[a]): Int32 =
+        Reducible.reduceMap(_ -> 1, t)
```

Comment:
I think foldLeft with + is more standard.

Reply:
Ill keep this example for a paper. Its brilliant(ly complicated!)

Reply:
It's "research", right? :wink: